crates/client/sync/src/commitments/lib.rs

```diff
@@ -198,29 +206,46 @@ fn contract_trie_root<B: BlockT>(
     csd: &CommitmentStateDiff,
     overrides: Arc<OverrideHandle<Block<Header<u32, BlakeTwo256>, OpaqueExtrinsic>>>,
     mut bonsai_contract: MutexGuard<BonsaiStorage<BasicId, BonsaiDb<B>, Pedersen>>,
+    bonsai_contract_storage: Arc<Mutex<BonsaiStorage<BasicId, BonsaiDb<B>, Pedersen>>>,
     block_number: u64,
     maybe_block_hash: Option<H256>,
 ) -> Result<Felt252Wrapper, BonsaiStorageError<BonsaiDbError>> {
+    let identifier = "0xcontract".as_bytes();
+    bonsai_contract.init_tree(&identifier)?;
+
+    // First we insert the contract storage changes
+    // TODO: @cchudant parallelize this loop
     for (contract_address, updates) in csd.storage_updates.iter() {
+        update_storage_trie(contract_address, updates, &bonsai_contract_storage);
+    }
+
+    // Then we commit them
+    bonsai_contract_storage.lock().unwrap().commit(BasicId::new(block_number))?;
+
+    // Then we compute the leaf hashes retrieving the corresponding storage root
+    // TODO: @cchudant parallelize this loop
+    for contract_address in csd.storage_updates.iter() {
         let class_commitment_leaf_hash =
-            contract_state_leaf_hash(csd, overrides.as_ref(), contract_address, updates, maybe_block_hash)?;
+            contract_state_leaf_hash(csd, &overrides, contract_address.0, maybe_block_hash, &bonsai_contract_storage)?;
 
-        let key = key(contract_address.0.0);
-        bonsai_contract.insert(&key, &class_commitment_leaf_hash.into())?;
+        let key = key(contract_address.0.0.0);
+        bonsai_contract.insert(&identifier, &key, &class_commitment_leaf_hash.into())?;
     }
 
     bonsai_contract.commit(BasicId::new(block_number))?;
-    Ok(bonsai_contract.root_hash()?.into())
+    Ok(bonsai_contract.root_hash(&identifier)?.into())
 }
 
-fn contract_state_leaf_hash(
+fn contract_state_leaf_hash<B: BlockT>(
     csd: &CommitmentStateDiff,
     overrides: &OverrideHandle<Block<Header<u32, BlakeTwo256>, OpaqueExtrinsic>>,
     contract_address: &ContractAddress,
-    storage_updates: &IndexMap<StorageKey, StarkFelt>,
     maybe_block_hash: Option<H256>,
+    bonsai_contract_storage: &Arc<Mutex<BonsaiStorage<BasicId, BonsaiDb<B>, Pedersen>>>,
 ) -> Result<Felt252Wrapper, BonsaiStorageError<BonsaiDbError>> {
-    let storage_root = update_storage_trie(overrides, contract_address, storage_updates, maybe_block_hash)?;
+    let identifier = contract_address.0.0.0.as_bytes_ref();
```

Comment:
Maybe consider refactoring this into an `identifier` function, similar to `key`?